python/lsst/daf/butler/datastores/s3Datastore.py

```diff
@@ -0,0 +1,736 @@
+            location computed from the template.
+        PermissionError
+            Raised when check if file exists at target location in S3 can not
+            be made because IAM user used lacks s3:GetObject or s3:ListBucket
+            permissions.
+        """
+        if not self.constraints.isAcceptable(ref):
+            # Raise rather than use boolean return value.
+            raise DatasetTypeNotSupportedError(f"Dataset {ref} has been rejected by this datastore via"
+                                               " configuration.")
+
+        if formatter is None:
+            formatter = self.formatterFactory.getFormatterClass(ref)
+
+        # ingest can occur from file->s3 and s3->s3 (source can be file or s3,
+        # target will always be s3). File has to exist at target location. Two
+        # Schemeless URIs are assumed to obey os.path rules. Equivalent to
+        # os.path.exists(fullPath) check in PosixDatastore.
+        srcUri = ButlerURI(path)
+        if srcUri.scheme == 'file' or not srcUri.scheme:
+            if not os.path.exists(srcUri.ospath):
+                raise FileNotFoundError(f"File at '{srcUri}' does not exist; note that paths to ingest are "
+                                        "assumed to be relative to self.root unless they are absolute.")
+        elif srcUri.scheme == 's3':
+            if not s3CheckFileExists(srcUri, client=self.client)[0]:
+                raise FileNotFoundError("File at '{}' does not exist; note that paths to ingest are "
+                                        "assumed to be relative to self.root unless they are absolute."
+                                        .format(srcUri))
+        else:
+            raise NotImplementedError(f"Scheme type {srcUri.scheme} not supported.")
+
+        # Transfer is generaly None when put calls ingest. In that case file is
+        # uploaded in put, or already in proper location, so source location
+        # must be inside repository. In other cases, created target location
+        # must be inside root and source file must be deleted when 'move'd.
+        if transfer is None:
+            rootUri = ButlerURI(self.root)
+            if srcUri.scheme == "file":
+                raise RuntimeError(f"'{srcUri}' is not inside repository root '{rootUri}'. "
+                                   "Ingesting local data to S3Datastore without upload "
+                                   "to S3 is not allowed.")
+            elif srcUri.scheme == "s3":
+                if not srcUri.path.startswith(rootUri.path):
+                    raise RuntimeError(f"'{srcUri}' is not inside repository root '{rootUri}'.")
+            p = pathlib.PurePosixPath(srcUri.relativeToPathRoot)
+            pathInStore = str(p.relative_to(rootUri.relativeToPathRoot))
+            tgtLocation = self.locationFactory.fromPath(pathInStore)
+        elif transfer == "move" or transfer == "copy":
+            if srcUri.scheme == "file":
```

Comment:
This should also support scheme-less (local files). Also put an else at the end as a sanity check.